File: WuchangLiveWallpaper/app/src/main/java/com/halot455/wuchangwallpaper/MainActivity.java
package com.halot455.wuchangwallpaper;

import android.app.*;
import android.app.WallpaperManager;
import android.content.*;
import android.graphics.Canvas;
import android.os.*;
import android.view.*;
import android.widget.*;
import android.graphics.Color;
import android.graphics.drawable.GradientDrawable;

public class MainActivity extends Activity {
    private HeroView hero;
    private String current;

    @Override public void onCreate(Bundle state) {
        super.onCreate(state);
        getWindow().setStatusBarColor(Color.BLACK);
        getWindow().setNavigationBarColor(Color.BLACK);
        current = getSharedPreferences("wuchang", MODE_PRIVATE).getString("character", "white");

        FrameLayout root = new FrameLayout(this);
        hero = new HeroView(this);
        root.addView(hero, new FrameLayout.LayoutParams(-1, -1));

        LinearLayout controls = new LinearLayout(this);
        controls.setOrientation(LinearLayout.VERTICAL);
        controls.setPadding(dp(16), dp(12), dp(16), dp(18));
        GradientDrawable bg = new GradientDrawable(GradientDrawable.Orientation.TOP_BOTTOM, new int[]{0x18000000, 0xCC05070D});
        controls.setBackground(bg);

        TextView title = new TextView(this);
        title.setText("VÔ THƯỜNG LIVE");
        title.setTextColor(Color.WHITE);
        title.setTextSize(19);
        title.setGravity(Gravity.CENTER);
        title.setPadding(0, 0, 0, dp(8));
        controls.addView(title, new LinearLayout.LayoutParams(-1, dp(44)));

        LinearLayout row = new LinearLayout(this);
        row.setOrientation(LinearLayout.HORIZONTAL);
        Button white = button("Bạch Vô Thường");
        Button black = button("Hắc Vô Thường");
        row.addView(white, new LinearLayout.LayoutParams(0, dp(52), 1));
        LinearLayout.LayoutParams sep = new LinearLayout.LayoutParams(dp(8), dp(1));
        Space sp = new Space(this); row.addView(sp, sep);
        row.addView(black, new LinearLayout.LayoutParams(0, dp(52), 1));
        controls.addView(row);

        Button set = button("ĐẶT LÀM HÌNH NỀN ĐỘNG");
        LinearLayout.LayoutParams setLp = new LinearLayout.LayoutParams(-1, dp(56));
        setLp.topMargin = dp(10);
        controls.addView(set, setLp);

        TextView hint = new TextView(this);
        hint.setText("Kéo trái/phải để xoay • tự động idle • tà váy/tóc bay nhẹ");
        hint.setTextColor(0xFFBBC3D8);
        hint.setTextSize(12);
        hint.setGravity(Gravity.CENTER);
        hint.setPadding(0, dp(7), 0, 0);
        controls.addView(hint, new LinearLayout.LayoutParams(-1, dp(36)));

        FrameLayout.LayoutParams cp = new FrameLayout.LayoutParams(-1, dp(218), Gravity.BOTTOM);
        root.addView(controls, cp);
        setContentView(root);

        white.setOnClickListener(v -> select("white"));
        black.setOnClickListener(v -> select("black"));
        set.setOnClickListener(v -> setWallpaper());
        select(current);
    }

    private void select(String c) {
        current = c;
        getSharedPreferences("wuchang", MODE_PRIVATE).edit().putString("character", c).apply();
        hero.character = c;
        hero.invalidate();
    }

    private Button button(String text) {
        Button b = new Button(this);
        b.setText(text);
        b.setTextColor(Color.WHITE);
        b.setTextSize(13);
        GradientDrawable gd = new GradientDrawable();
        gd.setColor(0xDD151C2D);
        gd.setCornerRadius(dp(14));
        gd.setStroke(dp(1), 0xFF3A4664);
        b.setBackground(gd);
        return b;
    }

    private void setWallpaper() {
        ComponentName cn = new ComponentName(this, WuchangWallpaperService.class);
        try {
            Intent i = new Intent(WallpaperManager.ACTION_CHANGE_LIVE_WALLPAPER);
            i.putExtra(WallpaperManager.EXTRA_LIVE_WALLPAPER_COMPONENT, cn);
            startActivity(i);
        } catch (Exception e) {
            startActivity(new Intent(WallpaperManager.ACTION_LIVE_WALLPAPER_CHOOSER));
        }
    }

    private int dp(int v) { return (int)(v * getResources().getDisplayMetrics().density + .5f); }

    final class HeroView extends View {
        final HeroRenderer renderer = new HeroRenderer(MainActivity.this);
        String character = current;
        float angle = 0f;
        float target = 0f;
        float lastX;
        boolean dragging;
        long prev = SystemClock.uptimeMillis();

        HeroView(Context c) { super(c); setLayerType(View.LAYER_TYPE_HARDWARE, null); }

        @Override protected void onDraw(Canvas canvas) {
            super.onDraw(canvas);
            long now = SystemClock.uptimeMillis();
            float dt = Math.min(.05f, (now - prev) / 1000f); prev = now;
            if (!dragging) target = (float)Math.sin(now * 0.00055f) * 26f;
            angle += (target - angle) * .09f;
            renderer.draw(canvas, getWidth(), getHeight(), angle, now, character, .5f);
            postInvalidateOnAnimation();
        }

        @Override public boolean onTouchEvent(android.view.MotionEvent e) {
            switch (e.getActionMasked()) {
                case MotionEvent.ACTION_DOWN: dragging = true; lastX = e.getX(); return true;
                case MotionEvent.ACTION_MOVE:
                    float dx = e.getX() - lastX; lastX = e.getX(); target = Math.max(-34f, Math.min(34f, target + dx * .22f)); return true;
                case MotionEvent.ACTION_CANCEL:
                case MotionEvent.ACTION_UP: dragging = false; return true;
            }
            return true;
        }
    }
}

File: WuchangLiveWallpaperKotlin/app/src/main/java/com/halot455/wuchangwallpaper/MainActivity.kt
package com.halot455.wuchangwallpaper

import android.app.Activity
import android.app.WallpaperManager
import android.content.ComponentName
import android.content.Intent
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.drawable.GradientDrawable
import android.os.Bundle
import android.os.SystemClock
import android.view.Gravity
import android.view.MotionEvent
import android.view.View
import android.widget.*

class MainActivity : Activity() {
    private lateinit var heroView: HeroView
    private var currentCharacter = "white"

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        window.statusBarColor = Color.BLACK
        window.navigationBarColor = Color.BLACK

        currentCharacter = getSharedPreferences("wuchang", MODE_PRIVATE)
            .getString("character", "white") ?: "white"

        val root = FrameLayout(this)
        heroView = HeroView()
        root.addView(heroView, FrameLayout.LayoutParams(-1, -1))

        val panel = LinearLayout(this).apply {
            orientation = LinearLayout.VERTICAL
            setPadding(dp(16), dp(12), dp(16), dp(18))
            background = GradientDrawable(
                GradientDrawable.Orientation.TOP_BOTTOM,
                intArrayOf(0x10000000, 0xE005070D.toInt())
            )
        }

        val title = TextView(this).apply {
            text = "VÔ THƯỜNG LIVE — KOTLIN"
            setTextColor(Color.WHITE)
            textSize = 18f
            gravity = Gravity.CENTER
        }
        panel.addView(title, LinearLayout.LayoutParams(-1, dp(42)))

        val row = LinearLayout(this).apply { orientation = LinearLayout.HORIZONTAL }
        val whiteButton = makeButton("Bạch Vô Thường")
        val blackButton = makeButton("Hắc Vô Thường")
        row.addView(whiteButton, LinearLayout.LayoutParams(0, dp(52), 1f))
        row.addView(Space(this), LinearLayout.LayoutParams(dp(8), 1))
        row.addView(blackButton, LinearLayout.LayoutParams(0, dp(52), 1f))
        panel.addView(row)

        val applyButton = makeButton("ĐẶT LÀM HÌNH NỀN ĐỘNG")
        val applyLp = LinearLayout.LayoutParams(-1, dp(56)).apply { topMargin = dp(10) }
        panel.addView(applyButton, applyLp)

        val hint = TextView(this).apply {
            text = "Kéo trái/phải để xoay • tự idle • tóc/tà áo lay nhẹ • không cần mạng"
            setTextColor(0xFFBBC3D8.toInt())
            textSize = 12f
            gravity = Gravity.CENTER
        }
        panel.addView(hint, LinearLayout.LayoutParams(-1, dp(34)))

        root.addView(panel, FrameLayout.LayoutParams(-1, dp(210), Gravity.BOTTOM))
        setContentView(root)

        whiteButton.setOnClickListener { selectCharacter("white") }
        blackButton.setOnClickListener { selectCharacter("black") }
        applyButton.setOnClickListener { openWallpaperPicker() }

        selectCharacter(currentCharacter)
    }

    private fun selectCharacter(value: String) {
        currentCharacter = value
        getSharedPreferences("wuchang", MODE_PRIVATE)
            .edit().putString("character", value).apply()
        heroView.character = value
        heroView.invalidate()
    }

    private fun openWallpaperPicker() {
        val component = ComponentName(this, WuchangWallpaperService::class.java)
        val intent = Intent(WallpaperManager.ACTION_CHANGE_LIVE_WALLPAPER).apply {
            putExtra(WallpaperManager.EXTRA_LIVE_WALLPAPER_COMPONENT, component)
        }
        try {
            startActivity(intent)
        } catch (_: Throwable) {
            startActivity(Intent(WallpaperManager.ACTION_LIVE_WALLPAPER_CHOOSER))
        }
    }

    private fun makeButton(label: String): Button {
        return Button(this).apply {
            text = label
            setTextColor(Color.WHITE)
            textSize = 13f
            isAllCaps = false
            background = GradientDrawable().apply {
                setColor(0xDD151C2D.toInt())
                cornerRadius = dp(14).toFloat()
                setStroke(dp(1), 0xFF3A4664.toInt())
            }
        }
    }

    private fun dp(v: Int): Int =
        (v * resources.displayMetrics.density + 0.5f).toInt()

    inner class HeroView : View(this) {
        private val renderer = HeroRenderer(this@MainActivity)
        var character: String = currentCharacter
        private var angle = 0f
        private var targetAngle = 0f
        private var lastX = 0f
        private var dragging = false

        override fun onDraw(canvas: Canvas) {
            super.onDraw(canvas)
            val now = SystemClock.uptimeMillis()
            if (!dragging) {
                targetAngle = kotlin.math.sin(now * 0.00055).toFloat() * 25f
            }
            angle += (targetAngle - angle) * 0.09f
            renderer.draw(canvas, width, height, angle, now, character, 0.5f)
            postInvalidateOnAnimation()
        }

        override fun onTouchEvent(event: MotionEvent): Boolean {
            when (event.actionMasked) {
                MotionEvent.ACTION_DOWN -> {
                    parent?.requestDisallowInterceptTouchEvent(true)
                    dragging = true
                    lastX = event.x
                    return true
                }
                MotionEvent.ACTION_MOVE -> {
                    val dx = event.x - lastX
                    lastX = event.x
                    targetAngle = (targetAngle + dx * 0.23f).coerceIn(-34f, 34f)
                    return true
                }
                MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
                    dragging = false
                    parent?.requestDisallowInterceptTouchEvent(false)
                    return true
                }
            }
            return true
        }
    }
}
